### ISSUE TITLE
First of all, thank you for sharing your code, it helps us a lot.   Can you show us the code for the statistical model activation in Figure 2 of the paper?

### ISSUE THREAD

**Issue by LiZhangMing** [yyyyychen/LowMemoryBP#7]
First of all, thank you for sharing your code, it helps us a lot.   Can you show us the code for the statistical model activation in Figure 2 of the paper?    llama13b is ok 
<img width="599" alt="image" src="https://github.com/user-attachments/assets/d15fe028-e445-4d1b-bd70-5c2a22991cbf">

**Comment by yyyyychen** [comment]
The ratios in the figure are not summarized from the practical code running. They are derived from our theoretical analysis of the memory mechanism. The detailed analysis for llama13b is shown in Figure 6.
 ![1732679013509](https://github.com/user-attachments/assets/8f5efc1b-e139-484f-a0c4-6423bb5341fb)

**Comment by LiZhangMing** [comment]
Thank you very much for your response. I have a small question regarding the activation memory generated during standard LoRA-based large model fine-tuning. Is the activation memory represented by the rightmost part of Figure 6? If so, in the LoRA formula \( Y = XW + XAB \), where \( W \) represents the original frozen weights and \( A \) and \( B \) are the low-rank trainable weights of LoRA, have you explicitly measured the activations such as \( XA \)?

**Comment by yyyyychen** [comment]
> Is the activation memory represented by the rightmost part of Figure 6?

Yes.

> If so, in the LoRA formula ( Y = XW + XAB ), where ( W ) represents the original frozen weights and ( A ) and ( B ) are the low-rank trainable weights of LoRA, have you explicitly measured the activations such as ( XA )?

No, we neglected the activation of XA. If A and B are both trainable, we have to save X for dA and XA for dB, but XA is usually of much smaller size than X. Since XA's shape is [B, N, lora_r], X's shape is [B, N, C] and lora_r is usually very small, the memory of them are not comparable.

**Comment by LiZhangMing** [comment]
<img width="840" alt="image" src="https://github.com/user-attachments/assets/11b469c6-379e-41ce-852f-7bd20de4db50">
Thank you very much for your response. In the image above, does the red-highlighted section represent the activation memory generated during LoRA fine-tuning of a large model with frozen original parameters? I'm not entirely sure about this. If so, then what scenario does the memory next to it represent?

**Comment by yyyyychen** [comment]
> In the image above, does the red-highlighted section represent the activation memory generated during LoRA fine-tuning of a large model with frozen original parameters?

Yes, it can represent the case that tuning all the linear layers by LoRA. In fact, it can also represent the full-tuning.

> If so, then what scenario does the memory next to it represent?

The next column is the case that all parameters in the transformer blocks (including lora_A and lora_B if using LoRA) are frozen. In real fine-tuning, we won't freeze all the parameters, but it is a good thought experiment to figure out what is the limit of activation memory that freezing can reduce.
